torchgen/model.py

```diff
@@ -1661,18 +2032,51 @@ def strip_arg_annotation(a: Argument) -> Argument:
             post_self_positional=tuple(
                 map(strip_arg_annotation, self.post_self_positional)
             ),
+            # Since TensorOptions are droped, the post_tensor_options_kwargs are
+            # converted to pre_tensor_options_kwargs
             pre_tensor_options_kwarg_only=tuple(
                 map(strip_arg_annotation, self.pre_tensor_options_kwarg_only)
-            ),
-            # NB: tensor_options guaranteed to not have any alias annotations
-            tensor_options=self.tensor_options,
-            post_tensor_options_kwarg_only=tuple(
-                map(strip_arg_annotation, self.post_tensor_options_kwarg_only)
-            ),
+            )
+            + tuple(map(strip_arg_annotation, self.post_tensor_options_kwarg_only)),
+            # TensorOptions are dropped in signature,
+            # so we can pair factory functions with their out= variants.
+            tensor_options=None,
+            post_tensor_options_kwarg_only=tuple(),
             # out arguments are dropped in signature
             out=(),
         )
 
+    def remove_self_annotation(self) -> "Arguments":
+        assert self.self_arg is not None
+        return Arguments(
+            pre_self_positional=self.pre_self_positional,
+            self_arg=SelfArgument(
+                Argument(
+                    name=self.self_arg.argument.name,
+                    type=self.self_arg.argument.type,
+                    default=self.self_arg.argument.default,
+                    annotation=None,
+                )
+            ),
+            post_self_positional=self.post_self_positional,
+            pre_tensor_options_kwarg_only=self.pre_tensor_options_kwarg_only,
+            tensor_options=self.tensor_options,
+            post_tensor_options_kwarg_only=self.post_tensor_options_kwarg_only,
+            out=self.out,
+        )
+
+    def with_out_args(self, outs: List[Argument]) -> "Arguments":
+        assert len(self.out) == 0
+        return Arguments(
+            pre_self_positional=self.pre_self_positional,
+            self_arg=self.self_arg,
+            post_self_positional=self.post_self_positional,
+            pre_tensor_options_kwarg_only=self.pre_tensor_options_kwarg_only,
+            tensor_options=self.tensor_options,
+            post_tensor_options_kwarg_only=self.post_tensor_options_kwarg_only,
+            out=tuple(outs),
```

Comment:
ah, the pain of purely functional programming lol

Reply:
I'd be OK with using `dataclasses.replace` to make this less painful (you'll need to shut up mypy about it, it won't be able to figure out the typing)

Reply:
oh good call